### PR TITLE
Simplify async certificate validation

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
Simplify async certificate validation. Currently, the `ValidateResultCallback` can be passed in to `doVerifyChain()` one of two different way:
* Via the optional `callback` argument
* Via the optional `ssl_extended_info` argument (by invoking `createValidateResultCallback()`)
This is a bit confusing as both arguments are optional but the callback is not. 

In addition, the handling of `ssl_extended_info` is slightly cumbersome. Since this argument is optional, all the code which attempts to set the `ClientValidationStatus` status on it is forced to wrap it with `if (ssl_extended_info)`. It is also quite easy to omit a call to set this status.

To remedy these issues, this PR removes requires the `ValidateResultCallback` to passed via `callback`. It also removes the `ssl_extended_info` completely, and adds `ClientValidationStatus` to `onCertValidationResult()` and `ValidationResults`. This allows the caller to set the  `ssl_extended_info` as needed.

Risk Level: Low, no behavior change
Testing: Existing unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

File: source/extensions/transport_sockets/tls/context_impl.cc [hunk 1]
```diff
@@ -505,15 +505,14 @@ ValidationResults ContextImpl::customVerifyCertChain(
     extended_socket_info->setCertificateValidationStatus(Ssl::ClientValidationStatus::NotValidated);
```
Comment (danzh2010) [inline source/extensions/transport_sockets/tls/context_impl.cc:505]: Either `extended_socket_info->setCertificateValidationStatus()` needs to be called after `doVerifyCertChain()` returns non-Pending status as well, or both places should defer calling the setter to the caller of this method.

**Review by danzh2010 — COMMENTED** [review]
Nice cleanup! I left a few nits.
And it makes code review much much easier to have E-M in the same repo!

File: source/extensions/transport_sockets/tls/context_impl.cc (hunk 1 above)
Comment (danzh2010) [reply source/extensions/transport_sockets/tls/context_impl.cc:505]: The fact that this is not breaking any test makes me wonder if anyone cares about the client validation status at all.

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.h [hunk 2]
```diff
@@ -108,11 +107,12 @@ class DefaultCertValidator : public CertValidator, Logger::Loggable<Logger::Id::
                                   const std::vector<SanMatcherPtr>& subject_alt_name_matchers);
 
 private:
-  bool verifyCertAndUpdateStatus(Ssl::SslExtendedSocketInfo* ssl_extended_info, X509* leaf_cert,
+  bool verifyCertAndUpdateStatus(Envoy::Ssl::ClientValidationStatus* detailed_status,
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.h]: Good point. Done.

File: source/extensions/transport_sockets/tls/context_impl.cc (hunk 1 above)
Comment (RyanTheOptimist) [reply source/extensions/transport_sockets/tls/context_impl.cc:505]: > The fact that this is not breaking any test makes me wonder if anyone cares about the client validation status at all.

Yeah, I wondered about this as well. The PlatformBridge validator also wasn't calling `setCertificateValidationStatus()` and everything passed. I wonder if people only use the status in the failure cases, and we don't have broad test coverage of the failure cases?

> Either extended_socket_info->setCertificateValidationStatus() needs to be called after doVerifyCertChain() returns non-Pending status as well

Good point. Done.

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc [hunk 3]
```diff
@@ -340,42 +334,40 @@ ValidationResults DefaultCertValidator::doVerifyCertChain(
...
+                                       Envoy::Ssl::ClientValidationStatus::Validated, tls_alert,
                                        error_details};
 }
 
-void DefaultCertValidator::onVerifyError(Ssl::SslExtendedSocketInfo* ssl_extended_info,
-                                         absl::string_view error) {
-  if (ssl_extended_info) {
-    ssl_extended_info->setCertificateValidationStatus(Envoy::Ssl::ClientValidationStatus::Failed);
-  }
+void DefaultCertValidator::onVerifyError(absl::string_view error) {
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc]: Sure, done.

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc [hunk 4]
```diff
@@ -230,14 +235,7 @@ bool DefaultCertValidator::verifyCertAndUpdateStatus(
...
 
-  if (ssl_extended_info) {
-    if (ssl_extended_info->certificateValidationStatus() ==
-        Envoy::Ssl::ClientValidationStatus::NotValidated) {
-      ssl_extended_info->setCertificateValidationStatus(validated);
-    } else if (validated != Envoy::Ssl::ClientValidationStatus::NotValidated) {
-      ssl_extended_info->setCertificateValidationStatus(validated);
-    }
-  }
+  *detailed_status = validated;
```
[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply source/extensions/transport_sockets/tls/cert_validator/default_validator.cc]: Oh, wow. I read the existing code several times as saying "if foo == NotValidated or foo != NotValidated". Thanks for catching this. (That being said, I'm not remotely sure I understand this logic, but c'est la vie :>). Done! Thanks!

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks, Dan!

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc [hunk 5]
```diff
@@ -340,46 +338,42 @@ ValidationResults DefaultCertValidator::doVerifyCertChain(
...
-      ssl_extended_info->setCertificateValidationStatus(
-          Envoy::Ssl::ClientValidationStatus::Validated);
-    }
   }
   std::string error_details;
   uint8_t tls_alert = SSL_AD_CERTIFICATE_UNKNOWN;
-  const bool succeeded = verifyCertAndUpdateStatus(
-      ssl_extended_info, leaf_cert, transport_socket_options.get(), &error_details, &tls_alert);
+  Envoy::Ssl::ClientValidationStatus detailed_status =
+      Envoy::Ssl::ClientValidationStatus::NotValidated;
```
Comment (danzh2010) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:366]: If verify_trusted_ca_  is true, the detailed status should start with`Envoy::Ssl::ClientValidationStatus:Validated` according to the removed line 359-362.

**Review by RyanTheOptimist — COMMENTED** [review]
Thanks!

**Review by danzh2010 — APPROVED** [review]
LGTM

**Comment by RyanTheOptimist** [comment]
FWIW, the clang-tidy issue is an annoying unfixable known issue :/

File: source/extensions/transport_sockets/tls/cert_validator/default_validator.cc [hunk 6]
```diff
@@ -340,46 +340,41 @@ ValidationResults DefaultCertValidator::doVerifyCertChain(
...
-      ssl_extended_info, leaf_cert, transport_socket_options.get(), &error_details, &tls_alert);
+  const bool succeeded = verifyCertAndUpdateStatus(leaf_cert, transport_socket_options.get(),
+                                                   detailed_status, &error_details, &tls_alert);
   return succeeded ? ValidationResults{ValidationResults::ValidationStatus::Successful,
-                                       absl::nullopt, absl::nullopt}
-                   : ValidationResults{ValidationResults::ValidationStatus::Failed, tls_alert,
+                                       Envoy::Ssl::ClientValidationStatus::Validated, absl::nullopt,
+                                       absl::nullopt}
+                   : ValidationResults{ValidationResults::ValidationStatus::Failed,
+                                       Envoy::Ssl::ClientValidationStatus::Validated, tls_alert,
```
Comment (ggreenway) [inline source/extensions/transport_sockets/tls/cert_validator/default_validator.cc:374]: Why is this one `Envoy::Ssl::ClientValidationStatus::Validated`? Should it be `detailed_status`?

**Review by ggreenway — CHANGES_REQUESTED** [review]
Nice cleanup!

/wait

File: envoy/ssl/ssl_socket_extended_info.h [hunk 7]
```diff
@@ -32,11 +32,14 @@ class ValidateResultCallback {
   /**
    * Called when the asynchronous cert validation completes.
    * @param succeeded true if the validation succeeds
+   * @param detailed_status detailed status of the underlying validation. Depending on the
+   *        validation configuration, `succeeded` may be true but `detailed_status` might
```
[the remark this answers is not in the document]
Comment (ggreenway) [reply envoy/ssl/ssl_socket_extended_info.h:36]: Oh! That rings a bell. I think it's for implementing this (from route_components.proto): 
```
  message TlsContextMatchOptions {
    option (udpa.annotations.versioning).previous_message_type =
        "envoy.api.v2.route.RouteMatch.TlsContextMatchOptions";

    // If specified, the route will match against whether or not a certificate is presented.
    // If not specified, certificate presentation status (true or false) will not be considered when route matching.
    google.protobuf.BoolValue presented = 1;

    // If specified, the route will match against whether or not a certificate is validated.
    // If not specified, certificate validation status (true or false) will not be considered when route matching.
    google.protobuf.BoolValue validated = 2;
  }
```

Ok, this makes some sense now (if I squint really hard). Can you add something to this effect to the comment?

[the remark this answers is not in the document]
Comment (RyanTheOptimist) [reply envoy/ssl/ssl_socket_extended_info.h:36]: Done. PTAL

**Review by ggreenway — APPROVED** [review]

**Pull request merged**